aarch64/kernel/memory_aarch64/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+
+#[cfg(any(target_arch = "aarch64", doc))]
+use {
+    core::arch::asm,
+    memory_structs::VirtualAddress,
+};
+
+const THESEUS_ASID: u16 = 0;
+
+#[cfg(any(target_arch = "aarch64", doc))]
+/// Flushes the specific virtual address in TLB.
+///
+/// TLBI => tlb invalidate instruction
+/// "va" => all translations at execution level
+///         using the supplied address
+/// "e1" => execution level
+pub fn tlb_flush_virt_addr(vaddr: VirtualAddress) {
+    #[cfg(target_arch = "aarch64")]
+    unsafe { asm!("tlbi vae1, {}", in(reg) vaddr.value()) };
+}
+
+#[cfg(any(target_arch = "aarch64", doc))]
+/// Flushes all TLB entries with Theseus' ASID (=0).
+///
+/// TLBI => tlb invalidate instruction
+/// "asid" => all entries with specific ASID
+/// "e1" => execution level
+pub fn tlb_flush_by_theseus_asid() {
+    #[cfg(target_arch = "aarch64")]
+    unsafe { asm!("tlbi aside1, {:x}", in(reg) THESEUS_ASID) };
+}
+
+#[cfg(any(target_arch = "aarch64", doc))]
+pub use tlb_flush_by_theseus_asid as tlb_flush_all;
+
+/// Returns the current top-level page table address.
+///
+/// We use TTBR0 in Theseus to store the
+/// top-level page table, so this function
+/// reads that register.
+pub fn get_p4() -> PhysicalAddress {
+    PhysicalAddress::new_canonical(
+        TTBR0_EL1.get_baddr() as usize
+    )
+}
+
+/// Disable the MMU using aarch64 registers
+///
+/// This uses the `SCTLR_EL1` register.
+///
+/// When the MMU is disabled, the CPU acts as
+/// if a full-address-space identity mapping
+/// was active.
+pub fn disable_mmu() {
+    SCTLR_EL1.modify(SCTLR_EL1::M::Disable);
+    unsafe { barrier::isb(barrier::SY) };
+}
+
+/// Enable the MMU using aarch64 registers
+///
+/// This uses the `SCTLR_EL1` register.
+///
+/// When the MMU is disabled, the CPU acts as
+/// if a full-address-space identity mapping
+/// was active. When it's enabled, the TTB0_EL1
+/// register is expected to point to a valid
+/// page table (using its physical address).
+pub fn enable_mmu() {
+    SCTLR_EL1.modify(SCTLR_EL1::M::Enable);
+    unsafe { barrier::isb(barrier::SY) };
+}
+
+/// Configures paging for Theseus.
+pub fn configure_translation_registers() {
+    unsafe {
+        // The MAIR register holds up to 8 memory profiles;
+        // each profile describes cacheability of the memory.
+        // In Theseus, we currently use two profiles: one for
+        // device memory (non-cacheable) and one for normal
+        // memory (the usual RAM memory, cacheable).
+        //
+        // For more information on MAIR, See section D17.2.97
+        // of [DDI0487l.a](https://l0.pm/arm-ddi0487l.a.pdf).
+        MAIR_EL1.write(
+            // Attribute 1 - Cacheable normal DRAM.
+            MAIR_EL1::Attr1_Normal_Outer::WriteBack_NonTransient_ReadWriteAlloc +
+            MAIR_EL1::Attr1_Normal_Inner::WriteBack_NonTransient_ReadWriteAlloc +
+
+            // Attribute 0 - Device.
+            MAIR_EL1::Attr0_Device::nonGathering_nonReordering_EarlyWriteAck,
```

Comment:
In `pte_flags_aarch64`, we expect the following:
* The [MAIR] index 0 has a Normal + Outer Shareable entry.
* The [MAIR] index 1 has a "DEVICE nGnRE" entry.

Perhaps I'm not understanding the code here, but it doesn't seem to match up. I don't see where the MAIR indices are specifically set.

Reply:
Wow I didn't notice that, you're totally right I'll swap them